src/main/java/com/tomtom/services/configuration/implementation/Configuration.java

```diff
@@ -653,19 +647,107 @@ private static boolean checkNodeMatchStringsChildren(@Nullable final List<NodeDT
     }
 
     /**
+     * Check that all parameters have a key and a value.
+     * @param params list of parameters to check
+     * @throws IncorrectConfigurationException If a parameter was found without key or value.
+     */
+    private static void checkAllParamsHaveKeyAndValue(final List<ParameterDTO> params) throws IncorrectConfigurationException {
+        for (final ParameterDTO p : params) {
+            if ((p.getKey() == null) ||
+                (p.getValue() == null)) {
+                throw new IncorrectConfigurationException("Parameter found without key or value!");
+            }
+        }
+    }
+
+    /**
+     * Expand all the includes in a parameter.
+     *
+     * @param tree     Parameter to expand.
+     * @param included Memory of which include files were processed.
+     * @throws IncorrectConfigurationException If include recursion was detected.
+     * @return Replacements for the parameter that was just expanded.
+     */
+    private static List<ParameterDTO> expandAllIncludesParams(
+            @Nonnull final ParameterDTO tree,
+            @Nonnull final List<String> included) throws IncorrectConfigurationException {
+        final String include = tree.getInclude();
+        final String includeArray = tree.getIncludeArray();
+        if (include != null) {
+            final List<ParameterDTO> replacement;
+            LOG.info("expandAllIncludesParams: Include specified, include={}", include);
+            // Check endless recursion.
+            if (included.contains(include)) {
+                throw new IncorrectConfigurationException("Endless recursion detected at include=" + include);
+            }
+
+            // Push name to stack.
+            included.add(0, include);
+
+            // Read JSON content from include.
+            final String content = readConfiguration(include);
+
+            // Parse nodes from content.
+            replacement = Arrays.asList(getChildObjectFromConfiguration(content, new TypeReference<ParameterDTO>(){}));
+
+            // Expand all includes in children, and construct list of replacements
+            final List<ParameterDTO> children = new ArrayList<>();
+            for (final ParameterDTO child : replacement) {
+                children.addAll(expandAllIncludesParams(child, included));
+            }
+
+            // Pop name from stack.
+            final String removed = included.remove(0);
+            assert removed.equals(include);
+            return children;
```

Comment:
Prefer to declare a `final List<ParameterDTO> result` at top of method and assign it here (and return it at end - easier for setting debug breakpoint).